Use the carbon component mapper for our DDF forms

skateman opened the pull request:
This replaces the old PF3 component mapper with the Carbon one, basically making all our DDF-based forms appear as Carbon forms. We have custom components that are appended to this mapper which should be addressed one by one:
* [x] code-editor (reimplemented)
* [x] field-array (provided by the [mapper](https://data-driven-forms.org/mappers/field-array?mapper=carbon))
* [x] dual-group (removed [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/7506))
* [x] dual-list-select (provided by the [mapper](https://data-driven-forms.org/mappers/dual-list-select?mapper=carbon))
* [x] password-field (UX?)
* [x] edit-password-field (UX?)
* [x] tree-view
* [x] tree-selector (UX?) 
* [x] select
* [x] protocol-selector (inherited from select)
* [x] validate-credentials
* [x] provider-credentials (no PF deps)
* [x] validate-provider-credentials (inherited from validate-credentials)
* [x] ~~detect-button (UX?)~~

There's also a styling issue which is probably a collision with our stylesheets pulled in from Carbon. The demo provided by DDF uses a CDN [link](https://github.com/data-driven-forms/react-forms/blob/master/packages/carbon-component-mapper/demo/index.html#L7) for the stylesheets so we might want to investigate the differences.

Resolves https://github.com/ManageIQ/manageiq-ui-classic/issues/7278

skateman commented:
Right now the form looks like this:
![Screenshot from 2020-10-29 13-54-58](https://user-images.githubusercontent.com/649130/97576365-5db53680-19ee-11eb-9c98-d69a06027805.png)

If I switch the theme to g10 in the `app/stylesheet/carbon.scss`, it looks a little bit better, but it alters the menu to white:
![Screenshot from 2020-10-29 13-50-34](https://user-images.githubusercontent.com/649130/97576456-7d4c5f00-19ee-11eb-84d3-49de01edcf4f.png)

I guess we have dark mode enabled?! @kavyanekkalapu @himdel

kavyanekkalapu commented:
Yes we should use g10 light mode  and we can fix leftnav colors manually.

skateman commented:
@kavyanekkalapu I started playing with the menu and its styling and I think we might be using the wrong one. There's a `bx--navigation` which is always dark independently from the theme and a `bx--side-nav` which follows the color scheme from the theme. Right now we're using the `bx--side-nav`, but is it possible that it's the wrong one?

himdel commented:
Ah, a different theme can be enabled for different parts of the app, by only `@import`ing the theme under a selector, instead of globally.

Failing that, we can trivially switch to a white theme, the menu can be de-themified by running `perl -i -npe 's/: \$.*\/\/ /: \$/ && s/$/;/' app/stylesheet/menu-colors.scss`.

skateman commented:
Since https://github.com/ManageIQ/manageiq-ui-classic/pull/7479 is in, we are now with the correct theme:
![Screenshot from 2020-11-06 13-06-41](https://user-images.githubusercontent.com/649130/98364439-19024e80-2031-11eb-9838-82cbeca0acea.png)

Isn't it beautiful? :heart_eyes_cat:

skateman commented:
Blocked by https://github.com/data-driven-forms/react-forms/pull/884

skateman commented:
### CodeEditor
Just changed the radio buttons and the group around them
![Screenshot from 2020-11-13 13-44-25](https://user-images.githubusercontent.com/649130/99073547-642edb00-25b6-11eb-9d49-638d620affaa.png)

### DualGroup
Switched to the implementation provided by the component mapper, that has a few caveats. The buttons don't have a tooltip setting, so I had to put text on them to be a11y compliant. There's a filter for both columns, which I find a little bit ugly and I might create an upstream PR that allows us to hide it. Also there's no way to hide some of the buttons as it was possible in the original implementation (which is IMO fine).
![Screenshot from 2020-11-13 13-48-30](https://user-images.githubusercontent.com/649130/99074071-65143c80-25b7-11eb-9941-d853567274b3.png)

### (Edit)PasswordField
There's no [input group (input addon)](https://getbootstrap.com/docs/4.0/components/input-group/) functionality in Carbon, so I don't know where to put the buttons for now. Asked UX to make my naive implementation a little bit better, waiting for their response.
![Screenshot from 2020-11-13 13-59-07](https://user-images.githubusercontent.com/649130/99074668-7c075e80-25b8-11eb-93f3-08e1c7c39846.png)
![Screenshot from 2020-11-13 13-59-16](https://user-images.githubusercontent.com/649130/99074664-7b6ec800-25b8-11eb-8a62-d82e32177a9d.png)

### AsyncCredentials
Not sure about the button style, need to talk with UX.
![Screenshot from 2020-11-13 14-17-35](https://user-images.githubusercontent.com/649130/99076158-ffc24a80-25ba-11eb-93ca-509bf40057d5.png)
![Screenshot from 2020-11-13 14-00-08](https://user-images.githubusercontent.com/649130/99074740-993c2d00-25b8-11eb-825b-4c5195b57d56.png)
![Screenshot from 2020-11-13 14-16-55](https://user-images.githubusercontent.com/649130/99076118-ecaf7a80-25ba-11eb-89e7-664a1394f9ca.png)

skateman commented:
Hangs on https://github.com/data-driven-forms/react-forms/pull/896

skateman commented:
UX got back to me with the input addons:
![image](https://user-images.githubusercontent.com/649130/101772754-155c5e80-3aec-11eb-8a5a-ef9cbd20b5be.png)

Unfortunately these aren't implemented as a react component, so we have to do it on our own in the DDF mapper.

https://github.com/data-driven-forms/react-forms/issues/912

skateman commented:
So the carbon team wasn't really helpful with the input addon, I decided to give up on this as I was explaining the problem over and over success. So instead of doing this *The Right Way&trade;*, I'm going to keep it as it is right now in our existing `PasswordField` and `EditPasswordField` components and just make the edit/cancel button inline with the UX suggested icon.

![Screenshot from 2021-01-18 14-28-05](https://user-images.githubusercontent.com/649130/104921301-6ab44700-5999-11eb-84e5-0381784e8190.png)
![Screenshot from 2021-01-18 14-28-10](https://user-images.githubusercontent.com/649130/104921297-69831a00-5999-11eb-9f8c-f9c0266604dc.png)

skateman commented:
TreeViewSelector doesn't exist in our codebase, so here's an example after the conversion:
![Screenshot from 2021-01-20 08-58-14](https://user-images.githubusercontent.com/649130/105144568-b7b12e00-5afd-11eb-98f3-d90243a0e466.png)
![Screenshot from 2021-01-20 08-58-21](https://user-images.githubusercontent.com/649130/105144559-b5e76a80-5afd-11eb-91bc-752510cae4f0.png)

It might need some tweaks when someone actually starts using it :wink:

skateman commented:
The (edit)-password-field has been also adjusted, thanks to @kavyanekkalapu 
![Screenshot from 2021-01-20 08-59-45](https://user-images.githubusercontent.com/649130/105144698-eaf3bd00-5afd-11eb-96bf-3a3134fa54ae.png)
![Screenshot from 2021-01-20 08-59-50](https://user-images.githubusercontent.com/649130/105144694-ea5b2680-5afd-11eb-9319-8384e6ff73e0.png)

kavyanekkalapu commented:
@skateman I noticed few things in the pr

- In carbon g10 theme, inputs and other fields in BG color light grey, check below screenshot.

<img width="1616" alt="Screen Shot 2021-01-20 at 9 37 00 AM" src="https://user-images.githubusercontent.com/37085529/105189895-7459be00-5b03-11eb-882d-f1acb7d99dd2.png">

MIQ form with no BG color on fields
<img width="1617" alt="Screen Shot 2021-01-20 at 9 40 36 AM" src="https://user-images.githubusercontent.com/37085529/105190056-9d7a4e80-5b03-11eb-8d02-c109b2e5068d.png">

- number input controls only appearing on hover or click which is very confusing , even width 100%, by looking at it it is not obvious that it is no input.
  
  Carbon No input
<img width="877" alt="Screen Shot 2021-01-20 at 9 22 16 AM" src="https://user-images.githubusercontent.com/37085529/105193305-7a9d6980-5b06-11eb-992e-dda89895be60.png">

MIQ no input( first field is also no input in screenshot, but it is not obvious)

<img width="1405" alt="Screen Shot 2021-01-20 at 9 15 57 AM" src="https://user-images.githubusercontent.com/37085529/105193517-930d8400-5b06-11eb-8474-2c408704de65.png">

skateman commented:
Okay, so summing up the problems you found:
* [ ] Theme is not loading correctly
* [ ] Incorrect component for the numeric input https://github.com/data-driven-forms/react-forms/issues/945
* [ ] Missing spacing between buttons (mapper?)
* [ ] `subform` title styling (mapper?)
* [ ] `dual-list` styling issues (mapper?)

himdel commented:
Aahh, I think I found it. The theme **is** loading properly and the colors are entirely consistent with Gray 10.

As per https://www.carbondesignsystem.com/components/form/style#color , the default input background is `$field-01`.
As per https://www.carbondesignsystem.com/guidelines/color/usage/, `$field-01` is only gray in the White theme, it's white in the Gray 10 theme.

Carbon documentation uses the white theme, so any screenshots from carbon docs don't apply, we're using a different theme.

himdel commented:
Ah, then I suppose the question is, should we go with the white theme instead, or change the background color.

Presumably there's a form class that does that though? If not, looks like the color is `$ui-background`

himdel commented:
```diff
diff --git a/app/stylesheet/carbon.scss b/app/stylesheet/carbon.scss
index 02b908287c..544085d254 100644
--- a/app/stylesheet/carbon.scss
+++ b/app/stylesheet/carbon.scss
@@ -14,7 +14,7 @@ $css--default-type: false;
 
 // Use the gray 90 theme
 @import '~@carbon/themes/scss/themes';
-$carbon--theme: $carbon--theme--g10;
+$carbon--theme: $carbon--theme--white;
 @include carbon--theme();
 
 @import '~carbon-components/scss/globals/scss/styles.scss';
```

skateman commented:
Okay, I (partially) addressed the issues, the dual-list should be fixed in the upstream package independently from this PR. I am looking for functional errors, like something not working that worked before so please do a sweep of the forms...

kavyanekkalapu commented:
@skateman i verified the branch, 
- When i enter something in code editor, i see this message, but my json has many syntax issues, it did not highlight them. 
<img width="1087" alt="Screen Shot 2021-02-01 at 9 03 14 AM" src="https://user-images.githubusercontent.com/37085529/106469060-c3d9ab80-646c-11eb-9adc-67d1db7faedd.png">

-   I think,  we need some spacing between options and editor
<img width="1093" alt="Screen Shot 2021-02-01 at 8 58 45 AM" src="https://user-images.githubusercontent.com/37085529/106469094-d227c780-646c-11eb-936d-5a9b606ac836.png">

- Editor should be in other BG color may be same like input text fields.

skateman commented:
> @skateman i verified the branch,
> 
> * When i enter something in code editor, i see this message, but my json has many syntax issues, it did not highlight them.
> 
> <img alt="Screen Shot 2021-02-01 at 9 03 14 AM" width="1087" src="https://user-images.githubusercontent.com/37085529/106469060-c3d9ab80-646c-11eb-9adc-67d1db7faedd.png">
> 

Because the code editor doesn't do error highlighting, only syntax coloring, you can verify that in the patternfly version on master.


> * I think,  we need some spacing between options and editor
> 
> <img alt="Screen Shot 2021-02-01 at 8 58 45 AM" width="1093" src="https://user-images.githubusercontent.com/37085529/106469094-d227c780-646c-11eb-936d-5a9b606ac836.png">
> 
> * Editor should be in other BG color may be same like input text fields.

This is a styling issue

h-kataria commented:
@skateman tested in UI, looking good to me other than styling issues. What is the next step?

skateman commented:
> @skateman tested in UI, looking good to me other than styling issues. What is the next step?

lol, I guess to merge? :rofl:

h-kataria commented:
> > @skateman tested in UI, looking good to me other than styling issues. What is the next step?
> 
> lol, I guess to merge? 🤣

wasn't sure if @kavyanekkalapu is still testing or found any other issues. @kavyanekkalapu are you still testing this or are you good with merge?

skateman commented:
Let's wait with this until @DavidResende0's CloudVolume PR gets merged

skateman commented:
Extracted the styling issues into a [separate issue](https://github.com/ManageIQ/manageiq-ui-classic/issues/7633) so this can be safely merged.
